src/server/io_stream.rs

```diff
@@ -0,0 +1,115 @@
+use hyper::rt::{Read, ReadBufCursor, Write};
+use std::io::{self, IoSlice};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+/// A wrapper around a stream that implements both `AsyncRead` and `AsyncWrite`.
+pub struct IoStream<T> {
+    stream: T,
```

Comment:
I think I may want to play around with this a bit before starting the feature work mentioned in the message of this pull request. I want to see how things perform if `stream` is wrapped in an `RwLock`. I haven't tried yet but I imagine benchmarking the echo server with println messages in `poll_read` and `poll_write` may tell a story. I also am considering boxing the stream.